src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPools.cs

```diff
@@ -31,14 +31,25 @@ public HttpConnectionPools(int maxConnectionsPerServer)
         {
             _maxConnectionsPerServer = maxConnectionsPerServer;
             _pools = new ConcurrentDictionary<HttpConnectionKey, HttpConnectionPool>();
-            _cleaningTimer = new Timer(s => ((HttpConnectionPools)s).RemoveStalePools(), this, CleanPoolTimeoutMilliseconds, CleanPoolTimeoutMilliseconds);
+            // Start out with the timer not running, since we have no pools.
+            _cleaningTimer = new Timer(s => ((HttpConnectionPools)s).RemoveStalePools(), this, Timeout.Infinite, Timeout.Infinite);
         }
 
         /// <summary>Gets a pool for the specified endpoint, adding one if none existed.</summary>
         /// <param name="key">The endpoint for the pool.</param>
         /// <returns>The retrieved pool.</returns>
-        public HttpConnectionPool GetOrAddPool(HttpConnectionKey key) =>
-            _pools.GetOrAdd(key, (k, s) => new HttpConnectionPool(s), _maxConnectionsPerServer);
+        public HttpConnectionPool GetOrAddPool(HttpConnectionKey key)
+        {
+            // If we currently don't have any pools, a new one will be added for this
+            // connection. We should then start reaping stale pools.
+            if (_pools.IsEmpty)
```

Comment:
@stephentoub I also thought about starting the timer in the callback for creating a new pool, but I think there is a subtle race there. See here: https://gist.github.com/steji113/8eab7dfea6d2a16c630815148db5c831

I had come up with a safe solution before I heard back from you on not using the lock on the get path. We need to ensure that the timer is started after the new pool has been added to the collection so that it is not subsequently stopped immediately.